hack/generator/pkg/codegen/pipeline_remove_embedded_resources.go

```diff
@@ -0,0 +1,681 @@
+	}
+
+	return visitor
+}
+
+func (e embeddedResourceRemover) NewResourceRemovalTypeWalker(visitor astmodel.TypeVisitor) *astmodel.TypeWalker {
+	typeWalker := astmodel.NewTypeWalker(e.types, visitor)
+	typeWalker.AfterVisit = func(original astmodel.TypeDefinition, updated astmodel.TypeDefinition, ctx interface{}) (astmodel.TypeDefinition, error) {
+		typedCtx := ctx.(resourceRemovalVisitorContext)
+
+		if !original.Name().Equals(updated.Name()) {
+			panic(fmt.Sprintf("Unexpected name mismatch during type walk: %q -> %q", original.Name(), updated.Name()))
+		}
+
+		if !original.Type().Equals(updated.Type()) {
+			flaggedType := e.typeFlag.ApplyTo(updated.Type())
+			var newName astmodel.TypeName
+			exists := false
+			for count := 0; ; count++ {
+				newName = embeddedResourceTypeName{original: original.Name(), context: typedCtx.resource.Name(), suffix: e.typeSuffix, count: count}.ToTypeName()
+				existing, ok := typedCtx.modifiedTypes[newName]
+				if !ok {
+					break
+				}
+				if existing.Type().Equals(flaggedType) {
+					exists = true
+					// Shape matches what we have already, can proceed
+					break
+				}
+			}
+			updated = updated.WithName(newName)
+			updated = updated.WithType(flaggedType)
+			if !exists {
+				typedCtx.modifiedTypes.Add(updated)
+			}
+
+			klog.V(5).Infof("Updating %q to %q", original.Name(), updated.Name())
+		}
+
+		return updated, nil
+	}
+
+	typeWalker.ShouldRemoveCycle = func(def astmodel.TypeDefinition, ctx interface{}) (bool, error) {
+		// If we're about to walk a cycle that is to a known resource type, just skip it entirely
+		if e.resourcePropertiesTypes.Contains(def.Name()) || e.resourceStatusTypes.Contains(def.Name()) {
+			return true, nil
+		}
+
+		if isTypeResourceLookalike(def.Type()) {
+			klog.V(5).Infof("Type %q is a resource lookalike", def.Name())
+			return true, nil
+		}
+
+		return false, nil // Leave other cycles for now
+	}
+
+	return typeWalker
+}
+
+func (e embeddedResourceRemover) removeEmbeddedResourceProperties() (astmodel.Types, error) {
+	result := make(astmodel.Types)
+
+	visitor := e.MakeEmbeddedResourceRemovalTypeVisitor()
+	typeWalker := e.NewResourceRemovalTypeWalker(visitor)
+
+	for _, def := range e.types {
+		if astmodel.IsResourceDefinition(def) {
+			// TODO: Bit awkward that we're modifying typewalker here... maybe should bring back initial ctx param?
+			typeWalker.MakeContext = func(it astmodel.TypeName, ctx interface{}) (interface{}, error) {
+				if ctx == nil {
+					return resourceRemovalVisitorContext{resource: def.Name(), depth: 0, modifiedTypes: make(astmodel.Types)}, nil
+				}
+				typedCtx := ctx.(resourceRemovalVisitorContext)
+				return typedCtx.WithMoreDepth(), nil
+			}
+
+			updatedTypes, err := typeWalker.Walk(def)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, newDef := range updatedTypes {
+				err := result.AddAllowDuplicates(newDef)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
```

Comment:
Extracting the `MakeContext` out of the loop doesn't work because `def` is used in the lamdba and is defined by the loop variable.

I like just pushing the `TypeWalker` construction into the loop though - that means I can pull setting this `def` stuff into `NewResourceRemovalTypeWalker` which is cleaner.

Reply:
Did the above